src/compiler/checker.ts

```diff
@@ -2663,7 +2746,170 @@ namespace ts {
+
+            if (cjsModuleType === undefined || cjsExportsType === undefined) {
+                return unknownType;
+            }
+
+            traverse(lastArg.body);
+
+            // Someone assigned to 'module.exports', so ignore other assignments
+            if (assignedModuleType) {
+                return assignedModuleType;
+            }
+
+            // Otherwise return the collected property assignments
+            if (exportAssignedProperties.length > 0) {
+                return createAnonymousType(lastArg.symbol,
+                        exportAssignedPropTable,
+                        /*callSignatures*/ [],
+                        /*constructSignatures*/ [],
+                        /*stringIndexType*/ undefined,
+                        /*numberIndexType*/ undefined);
+            }
+
+            // Empty module?
+            return emptyObjectType;
+
+            function traverse(node: Node) {
+                if (node.kind === SyntaxKind.BinaryExpression &&
+                   (<BinaryExpression>node).operatorToken.kind === SyntaxKind.EqualsToken) {
+                    let lhs = (<BinaryExpression>node).left;
+                    if (lhs.kind === SyntaxKind.PropertyAccessExpression) {
+                        // e.g. module.exports = foo;
+                        let propertyName = (<PropertyAccessExpression>lhs).name.text;
+                        let operandType = getTypeOfNode((<PropertyAccessExpression>lhs).expression);
+                        if (operandType === cjsModuleType && propertyName === "exports") {
+                            assignedModuleType = getTypeOfNode((<BinaryExpression>node).right);
+                        }
+                        else if (operandType === cjsExportsType) {
+                            // e.g. exports = bar;
+                            let prop = <TransientSymbol>createSymbol(SymbolFlags.Property | SymbolFlags.Transient, propertyName);
+                            prop.type = getTypeOfNode((<BinaryExpression>node).right);
+                            exportAssignedProperties.push(prop);
+                            exportAssignedPropTable[propertyName] = prop;
+                        }
+                    }
+                }
+                forEachChild(node, traverse);
+            }
+        }
+
+        function getTypeOfDefineModule(symbol: Symbol): Type {
+            let links = getSymbolLinks(symbol);
+            if (!links.type) {
+                if (!pushTypeResolution(symbol, TypeSystemPropertyName.Type)) {
+                    return unknownType;
+                }
+
+                links.type = resolveDefineModule(symbol);
+            }
+            return links.type;
+        }
+
+        function getTypeOfAmdExportAssignment(symbol: Symbol): Type {
+            let links = getSymbolLinks(symbol);
+            if (!links.type) {
+                if (!pushTypeResolution(symbol, TypeSystemPropertyName.Type)) {

```

Comment:
Every call to `pushTypeResolution` needs a balancing call to `popTypeResolution`!